### ISSUE TITLE
research dispatching long-running CI jobs to pre-configured runner

### ISSUE THREAD

**Issue by cfm** [freedomofpress/securedrop#6074]
**Tl;dr:**  Don't let's reinvent CircleCI.  :-)

## Description

Currently all CI jobs, including the long-running `app-tests` and `staging-tests-with-rebase`, run on CircleCI's on-demand infrastructure.  That is, when a workflow is triggered, CircleCI spins up a container (`docker`) or VM (`machine`) with the required image, and off we go.  CircleCI has summarized the [tradeoffs between these two execution models](https://circleci.com/docs/2.0/executor-types/#docker-benefits-and-limitations).

Since a significant portion of SecureDrop's CI time goes to per-job (and, for parallelized jobs, per-run) setup, I've wondered if it would be possible to use a fixed [CircleCI runner](https://circleci.com/docs/2.0/runner-overview/).  Specifically, this runner would:

1. set up the test VM proactively;
2. be available to start in on (e.g.) `app-tests` jobs immediately, without taking any time for further setup; and
3. destroy the test VM after each run; `GOTO 1`.

In researching (and [diagramming](https://gist.github.com/cfm/37f672c091d9ee1d7ad5f077e8582afa)) this approach, I've identified a host of problems with it that it make not worth the trouble:

1. [**The CircleCI Runner is available only on the "Scale" plan.**](https://circleci.com/docs/2.0/runner-overview/)  Working around this limitation would replicate most of the complexity of `staging-test-with-rebase`, just with the goal of dispatching tests to the pre-configured VM rather than to one capable of nested virtualization.  In particular...
2. As with the nested-virtualization image, **CI would risk confusing developers** by behaving differently than might be defined in the ref under test.
3. **This custom runner would be able to handle jobs only serially,** unless we were to invest in dynamic orchestration of multiple instances.  Otherwise, either (a) performance would be *worse* than the status quo or (b) only the first job at any given would benefit from the custom runner, while jobs `2..n` would revert to the status quo.
4. **We'd lose the benefit of parallelization *within* jobs,** also without dynamic orchestration.

## User Stories

As a developer or contributor, I would like CI to get better, not worse.  ;-)

**Comment by cfm** [comment]
Logging for posterity; closing as unworkable.
